javascript-review_Moss.js

```diff
@@ -0,0 +1,293 @@
+  // 20.
+  // a.
+  // Uncomment the code below.
+  // let newNum = 1
+  // const myFunc = (num, callback) => {
+  //   newNum *= num
+  //   callback(newNum)
+  // }
+  // myFunc(10, sum => {
+  //   myFunc(10, sum => {
+  //     myFunc(10, sum => {
+  //       console.log(sum)
+  //     })
+  //   })
+  // })
+  // b.
+  // Explain in your own words how this code works. For example you could start with something like:
+  
+  // "Firstly, newNum is assign by 1 then the myFunc is initialized . Secondly, the myFunc function is invoked. It received the arguments of 10 and it call back function then newNum = newNum * the num pass and argument which is 10 then a callback function is run taking that newNum and so on (the second call Myfunc is run again)"
+  
+  // 21.
+  
+  // a. Uncomment the code from below
+  
+  // // this is a factory function, it's another syntax for creating objects similar to the constructor function syntax
+  
+  // function Person(name) {
+  //   function sayHi() {
+  //     console.log('hello')
+  //   }
+  //   return {
+  //     name,
+  //     sayHi
+  //   }
+  // }
+  
+  // const harrison = Person('harrison')
+  
+  // b.
+  // What does this function return?
+  // { name: 'harrison', sayHi: [Function: sayHi] }
+  
+  // c.
+  // How do you call the sayHi() function?
+  // harrison.sayHi()
+  
+  // 22.
+  // a.
+  // Using the code from above as a template, implement the same functionality using the class sugar syntax
+  // Note that you'll need to comment out the code from above for this to work
+  
+  class Person {
+    constructor(name) {
+      this.name = name
+      this.sayHi = this.sayHi
+    }
+    sayHi() {
+      console.log('hello')
+    }
+  }
+  
+  // const harrison = new Person('harrison')
+  // console.log(harrison)
+  // harrison.sayHi()
+  
+  // b.
+  // Implement another instance method for your class that takes in age and height (in cms) as arguments (both number type) and adds these arguments as attributes to your person object
+  
+  // console.log the updated person object showing all three attributes (name, age, height) as being a part of the person object
+  
+  class Person2 extends Person {
+    constructor(name, age, height) {
+      super(name)
+      this.age = age
+      this.height = height
+    }
+  }
+  const harrison = new Person2('harrison', 10, 2)
+  // // console.log(harrison)
+  // harrison.sayHi()
+  // 23.
+  
+  // Define a function named waitBeforeSum that takes 2 numbers as arguments. Your function should sum these numbers together but only after waiting for 4 seconds inside of a setTimeout.
+  
+  function waitBeforeSum(a, b) {
+    return new Promise(() => {
+      setTimeout(() => {
+        return a + b
+      }, 4)
+    })
+  }
+  
+  // What do we need to use to access the value in the setTimeout only after the 4 seconds? Write the code to achieve this.
+  // console.log(waitBeforeSum(1, 2))
+  
+  cosnt q23 = () => {
+  console.log(await waitBeforeSum(1, 2))
```

Comment:
careful with using the `await` keyword here, you'd need to make it an async function